### PR TITLE
amqp retry logic

### PR DESCRIPTION
**Pull request by saahildhulla**
The AMQP library in Forge does not do automatic reconnection, so we need to implement it. 

This works to add retry logic to two two cases where RMQ disconnects: 
1) on initial connection to the `publisher`
2) when publishing a message to RMQ

Unit tests need to be fixed

### CODE REVIEW

File: internal/buildjob/buildjob.go [hunk 1]
```diff
@@ -61,17 +61,16 @@ func New(cfg Config) (*Job, error) {
...
+	var publisher message.Publisher
 	// setup message publisher
-	var producer message.Producer
 	if cfg.BrokerOpts != nil {
 		log.Info("Initializing status update message publisher")
 
-		if producer, err = message.NewProducer(cfg.BrokerOpts); err != nil {
-			return nil, err
-		}
+		publisher, _ := message.NewPublisher(cfg.BrokerOpts)
```
Comment (saahildhulla) [inline internal/buildjob/buildjob.go:69]: need to figure out a good way to handle an error case, if any

File: internal/message/amqp/publisher_test.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package amqp
+
+import "testing"
```
Comment (saahildhulla) [inline internal/message/amqp/publisher_test.go:3]: test skeleton to be filled in

File: internal/message/message.go [hunk 3]
```diff
@@ -15,17 +15,17 @@ const AmqpBroker Broker = "amqp"
 // SupportedBrokers defines the list of implemented message publishers.
 var SupportedBrokers = []Broker{AmqpBroker}
 
-// Producer defines the operations required by all message producers.
-type Producer interface {
-	Publish(message interface{}) error
+// Publisher defines the operations required by all message producers.
+type Publisher interface {
```
Comment (saahildhulla) [inline internal/message/message.go:19]: Publisher seemed to make sense as a name, though open to suggestions if this is misleading @sonnysideup

File: internal/buildjob/buildjob.go [hunk 4]
```diff
@@ -61,17 +61,16 @@ func New(cfg Config) (*Job, error) {
...
+	var publisher message.Publisher
 	// setup message publisher
-	var producer message.Producer
 	if cfg.BrokerOpts != nil {
 		log.Info("Initializing status update message publisher")
 
-		if producer, err = message.NewProducer(cfg.BrokerOpts); err != nil {
-			return nil, err
-		}
+		publisher, _ := message.NewPublisher(cfg.BrokerOpts, log)
```
Comment (saahildhulla) [inline internal/buildjob/buildjob.go:69]: should we pass the same logger through to the publisher?

File: internal/buildjob/buildjob.go [hunk 5]
```diff
@@ -61,17 +61,19 @@ func New(cfg Config) (*Job, error) {
...
+	var publisher message.Publisher
 	// setup message publisher
-	var producer message.Producer
+	log.Info("Set up message publisher if options are configured")
 	if cfg.BrokerOpts != nil {
 		log.Info("Initializing status update message publisher")
 
-		if producer, err = message.NewProducer(cfg.BrokerOpts); err != nil {
-			return nil, err
+		if publisher = message.NewPublisher(cfg.BrokerOpts, log); publisher == nil {
```
Comment (saahildhulla) [inline internal/buildjob/buildjob.go:70]: error checking handled a bit differently from before, in that the Publisher constructor doesn't return an error as a default case

File: internal/message/amqp/publisher.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	return &p
+}
+
+func (p Publisher) handleReconnect(uri string) {
+	for {
+		p.isConnected = false
+		p.logger.Info("Attempting to connect")
+		for !p.connect(uri) {
+			p.logger.Info("Failed to connect. Retrying.")
+			time.Sleep(reconnectDelay)
```
Comment (saahildhulla) [inline internal/message/amqp/publisher.go:46]: not sure if Sleep is the right move here, even if it's for a short period of time (few seconds)

File: internal/message/amqp/publisher.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+	"errors"
+	"github.com/go-logr/logr"
+	"time"
+
+	"github.com/streadway/amqp"
+)
+
+const (
+	reconnectDelay = 5 * time.Second
+	resendDelay = 5 * time.Second
```
Comment (saahildhulla) [inline internal/message/amqp/publisher.go:14]: is 5 seconds too long?

File: internal/message/amqp/publisher.go [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+		name:          queueName,
+		logger:        log,
+		done:          make(chan bool),
+	}
+	go p.handleReconnect(uri)
+	return &p
+}
+
+func (p Publisher) handleReconnect(uri string) {
+	for {
```
Comment (saahildhulla) [inline internal/message/amqp/publisher.go:41]: was between this and a set number of connection attempts -- thoughts?

File: internal/message/amqp/publisher.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		done:          make(chan bool),
+	}
+	go p.handleReconnect(uri)
+	return &p
+}
+
+func (p *Publisher) handleReconnect(uri string) {
+	for {
+		p.isConnected = false
+		for !p.connect(uri) {
```
Comment (saahildhulla) [inline internal/message/amqp/publisher.go:43]: should we add a max number of attempts?

File: internal/message/amqp/publisher.go [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+	p.notifyConfirm = make(chan amqp.Confirmation)
+	p.channel.NotifyClose(p.notifyClose)
+	p.channel.NotifyPublish(p.notifyConfirm)
+}
+
+// will push data onto the queue and wait for a confirm.
+// it continuously resends messages until a confirm is received.
+func (p *Publisher) Push(event interface{}) error {
+
+	for {
```
Comment (saahildhulla) [inline internal/message/amqp/publisher.go:113]: should we add a max number of attempts?

File: internal/message/message.go (hunk 3 above)
Comment (sonnysideup) [reply internal/message/message.go:18]: Seems fine to me and aligns with the [rabbitmq nomenclature](https://www.rabbitmq.com/publishers.html). 👍

File: internal/message/amqp/publisher.go (hunk 9 above)
Comment (saahildhulla) [reply internal/message/amqp/publisher.go:43]: @sonnysideup any thoughts regarding retry limits are appreciated!

File: internal/message/amqp/publisher.go (hunk 10 above)
Comment (saahildhulla) [reply internal/message/amqp/publisher.go:113]: @sonnysideup any thoughts regarding retry limits are appreciated!

**Comment by saahildhulla** [comment]
```	We want two things out of an RMQ publisher:

	* Connection logic with some cap on the maximum number of attempts and a fixed delay between retries
	* The ability to signal a reconnect when an existing connection is closed

	There are 4 different operations that can fail whenever RMQ goes away:

	1. Dialing a server (i.e. create connection)
	2. Creating a channel
	3. Declaring a queue
	4. Publishing a message

	Important Design Questions:

	- Should the publisher house all of the logic required to connect/reconnect a Connection and/or Channel?

	This is a bigger design and might not be worth the logic separation in terms of CBA.

	- Should "connection lost" errors be auto-resolved or should that signal be passed along to a user?

	We should be able to auto-resolve our own connection lost errors, but the real question is "when do we resolve them?"

	- Should we watch connections and channels separately, or focus primarily on the root cnxn object?

		How can channels close independently of connections w/o an explicit Close() operation? Seems low-risk.

		Connection failures result in channel failures: this is a reliable failure mode.```

**Comment by sonnysideup** [comment]
@saahildhulla closing this now that #57 has been merged with the intended changes.

**Pull request closed without merging**